From 9c4166b0534a1f02081beda9ee7b49c70a449571 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Thu, 14 Aug 2025 11:10:11 +0800
Subject: [PATCH 111/123] mtk: mt76: mt7996: add debugfs to disable beacon loss
 disconnection

Usage:
echo 1 > /sys/kernel/debug/ieee80211/phy0/mt76/beacon_loss_keep_conn

STA will trigger disconnection when beacon loss happens and no ack
reported from QoS NULL probe.
Add a debugfs to disable the disconnection for debug purpose, for
example, checking which queues are hanged.

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 mt7996/mac.c         | 3 ++-
 mt7996/mt7996.h      | 2 ++
 mt7996/mtk_debugfs.c | 2 ++
 3 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index 2b91eab7..12447f78 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -3775,7 +3775,8 @@ void mt7996_beacon_mon_work(struct work_struct *work)
 			loss_duration = msecs_to_jiffies(MT7996_MAX_PROBE_TIMEOUT);
 			timeout = mvif->probe_send_time[band_idx] + loss_duration;
 			if (time_after_eq(jiffies, timeout)) {
-				if (mvif->probe_send_count[band_idx] == MT7996_MAX_PROBE_TRIES)
+				if (mvif->probe_send_count[band_idx] == MT7996_MAX_PROBE_TRIES &&
+				    !dev->dbg.beacon_loss_keep_conn)
 					state = MON_STATE_LINK_LOST;
 				else
 					state = MON_STATE_SEND_PROBE;
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 50205792..d7265ef9 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -982,6 +982,8 @@ struct mt7996_dev {
 			u8 min_tx_tpo;
 			u8 pst;
 		} lp;
+
+		bool beacon_loss_keep_conn;
 	} dbg;
 	const struct mt7996_dbg_reg_desc *dbg_reg;
 	bool sr_pp_enable:1;
diff --git a/mt7996/mtk_debugfs.c b/mt7996/mtk_debugfs.c
index ebd5b21f..b2b9a05c 100644
--- a/mt7996/mtk_debugfs.c
+++ b/mt7996/mtk_debugfs.c
@@ -4761,6 +4761,8 @@ void mt7996_mtk_init_dev_debugfs(struct mt7996_dev *dev, struct dentry *dir)
 #ifdef CONFIG_MTK_DEBUG
 	dev->dbg.sku_disable = true; /* For SQC */
 	debugfs_create_u8("sku_disable", 0600, dir, &dev->dbg.sku_disable);
+	debugfs_create_bool("beacon_loss_keep_conn", 0600, dir,
+			    &dev->dbg.beacon_loss_keep_conn);
 #endif
 
 	debugfs_create_file("muru_prot_thr", 0200, dir, dev, &fops_muru_prot_thr);
-- 
2.45.2

